File: src/main/java/com/example/backend/helpers/TarotHelper.java
package com.example.backend.helpers;

import com.example.backend.dto.tarot.TarotRequest;
import com.example.backend.dto.tarot.TarotResponse;
import com.example.backend.gpt.ChatGPTService;
import com.example.backend.lang.Language;
import com.example.backend.lang.YandexTranslateHelper;
import lombok.AllArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;

@Component
@AllArgsConstructor
@Slf4j
public class TarotHelper {

    private final ChatGPTService chatGPTHelper;
    private final YandexTranslateHelper yandexTranslateHelper;

    @Transactional
    public TarotResponse futureTell(TarotRequest request) {
        log.info("Try get prediction");
        TarotRequest translatedInEngRequest = translateRequestIfNecessary(request);
        TarotResponse tarotResponse = chatGPTHelper.tarotMeChatGPT(translatedInEngRequest);
        log.info("Successfully got tarot prediction");
        return translateResponseIfNecessary(request, tarotResponse);
    }

    private TarotRequest translateRequestIfNecessary(TarotRequest request) {
        if (request.from() != Language.EN) {
            log.info("Tarot request should be translated from={} to{}", request.from(), request.to());
            String questionInEnglish = yandexTranslateHelper.translate(request.text(), request.from(), request.to());
            return new TarotRequest(request.cards(), questionInEnglish, request.from(), request.to());
        }
        return request;
    }

    private TarotResponse translateResponseIfNecessary(TarotRequest request, TarotResponse tarotResponse) {
        if (request.from() != Language.EN) {
            log.info("Tarot response should be translated back from={} to{}", request.to(), request.from());
            String responseInEnglish = yandexTranslateHelper.translate(tarotResponse.text(), request.to(), request.from());
            return new TarotResponse(tarotResponse.cards(), responseInEnglish, request.to(), request.from());
        }
        return tarotResponse;
    }
}

File: src/main/kotlin/com/example/backend/helper/TarotHelper.kt
package com.example.backend.helper

import com.example.backend.dto.tarot.TarotRequest
import com.example.backend.dto.tarot.TarotResponse
import com.example.backend.gpt.ChatGPTService
import com.example.backend.lang.Language
import com.example.backend.lang.YandexTranslateHelper
import com.example.backend.utils.LOGGER
import org.springframework.stereotype.Component
import reactor.core.publisher.Mono

@Component
class TarotHelper(
    val chatGPTService: ChatGPTService,
    val yandexTranslateHelper: YandexTranslateHelper
) {

    fun futureTell(request: TarotRequest): Mono<TarotResponse?> {
        LOGGER.info { "Try get prediction" }
        val translatedInEngRequest = translateRequestIfNecessary(request)
        val tarotResponse: Mono<TarotResponse?> = chatGPTService.tarotMeChatGPT(translatedInEngRequest)
        val translatedResponse = translateResponseIfNecessary(request, tarotResponse)
        return translatedResponse
    }

    private fun translateResponseIfNecessary(
        request: TarotRequest,
        tarotResponse: Mono<TarotResponse?>
    ): Mono<TarotResponse?> {
        if (request.from != Language.EN) {
            LOGGER.info { "Tarot response should be translated back from=${request.to} to=${request.from}" }

            return tarotResponse.flatMap { response ->
                yandexTranslateHelper.translate(response?.text!!, request.to!!, request.from!!)
                    .map { translatedText ->
                        TarotResponse(response.cards, translatedText)
                    }
            }
        }
        return tarotResponse
    }

    private fun translateRequestIfNecessary(request: TarotRequest): Mono<TarotRequest> {
        if (request.from != Language.EN) {
            LOGGER.info { "Tarot request should be translated from=${request.from} to=${request.to}" }
            return yandexTranslateHelper.translate(request.text!!, request.from!!, request.to!!)
                .map { translatedQuestion ->
                    TarotRequest(request.cards, translatedQuestion, request.from, request.to)
                }
        }
        return Mono.just(request)
    }
}